File: lab_2/race.py
import random

fin = open('race.in', 'r')
fout = open('race.out', 'w')

data = fin.read().strip().split('\n')
people = dict()

for i in data[1:]:
    info = i.split()
    if info[0] in people:
        people[info[0]].append(info[1])
    else:
        people[info[0]] = [info[1]]

def quicksort(array: list) -> list:
    if len(array) < 2:
        return array
    else:
        index = random.randint(0, len(array) - 1)
        pivot = array[index]
        array.pop(index)
        less = [i for i in array if i <= pivot]
        greater = [i for i in array if i > pivot]

        result = quicksort(less) + [pivot] + quicksort(greater)
        return result

countries = quicksort(list(people.keys()))

string = ''
for j in countries:
    if string == '':
        string += ('=== ' + j + ' ===\n')
    else:
        string += ('\n=== ' + j + ' ===\n')
    string += '\n'.join(people[j])

print(string, file = fout)

File: lab_2/race.go
package main

import (
	"fmt"
	"io/ioutil"
	"math/rand"
	"os"
	"strings"
)

func quicksort(array [][]string) [][]string {
	n := len(array)
	if n < 2 {
		return array
	} else {
		index := rand.Intn(n)
		pivot := array[index]
		var less [][]string
		var greater [][]string

		for i := 0; i < n; i++ {
			if i != index {
				if array[i][0] < pivot[0] {
					less = append(less, array[i])
				} else if array[i][0] > pivot[0] {
					greater = append(greater, array[i])
				} else {
					if i < index {
						less = append(less, array[i])
					} else {
						greater = append(greater, array[i])
					}
				}
			}
		}

		result := append(quicksort(less), pivot)
		result = append(result, quicksort(greater)...)
		return result
	}
}

func main() {
	var n int
	data_raw, _ := ioutil.ReadFile("race.in")
	data := strings.Split(string(data_raw), "\n")
	fmt.Sscanf(data[0], "%d", &n)
	array := make([][]string, n)

	for i := 1; i <= n; i++ {
		array[i-1] = (strings.Split(data[i], " "))
	}

	result := quicksort(array)
	var result_final []string

	country := ""
	var names []string

	for i := 0; i < n; i++ {
		if country != result[i][0] {
			if country != "" {
				result_final = append(result_final, "=== "+country+" ==="+"\n"+strings.Join(names, "\n"))
			}
			country = result[i][0]
			names = nil
		}
		names = append(names, result[i][1])
	}
	if country != "" {
		result_final = append(result_final, "=== "+country+" ==="+"\n"+strings.Join(names, "\n"))
	}

	fout, _ := os.Create("race.out")
	fout.WriteString(strings.Join(result_final, "\n"))
	fout.Close()
}
